File: C++/109.cpp
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

/*有序列表转换为高度二叉搜索树
使用二分法将列表转换为二叉搜索树，类似于108题
*/
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        vector<int> nums;
        while (head != NULL) {
            nums.push_back(head->val);
            head = head->next;
        }
        return func(nums, 0, nums.size()-1);
    }

    TreeNode *func(vector<int> &nums, int left, int right) {
        if (left > right)
            return NULL;
        int rootIndex = (left+right)/2;
        int rootVal = nums.at(rootIndex);
        TreeNode *root = new TreeNode(rootVal);
        root->left = func(nums, left, rootIndex-1);
        root->right = func(nums, rootIndex+1, right);
        return root;
    }
};

File: Python/109.py
from typing import List
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        nums = []
        while head is not None:
            nums.append(head.val)
            head = head.next
        return self.func(nums, 0, len(nums)-1)

    def func(self, nums: List[int], left: int, right: int) -> TreeNode:
        if left > right:
            return None
        rootIndex = (left+right)//2
        root = TreeNode(nums[rootIndex])
        root.left = self.func(nums, left, rootIndex-1)
        root.right = self.func(nums, rootIndex+1, right)
        return root